# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/webrender_bindings/RenderCompositorD3D11SWGL.cpp
# Commit: f0d21589e3fd
# Full Hash: f0d21589e3fd67d7939e694a2f49adbaa7f5c6d6
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2020-11-10 09:57:49
# Description:
#   Bug 1675665 - Don't draw surfaces if we're not currently presenting a frame. r=lsalzman
#   
#   WR can run an update which outputs the compositor surfaces, but never calls Begin/EndFrame.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D96380
# ==============================================================================

diff -r af637d06295a -r f0d21589e3fd gfx/webrender_bindings/RenderCompositorD3D11SWGL.cpp
--- a/gfx/webrender_bindings/RenderCompositorD3D11SWGL.cpp	Mon Nov 09 21:35:50 2020 +0000
+++ b/gfx/webrender_bindings/RenderCompositorD3D11SWGL.cpp	Mon Nov 09 22:05:46 2020 +0000
@@ -58,20 +58,31 @@
 }
 
 bool RenderCompositorD3D11SWGL::BeginFrame() {
+  MOZ_ASSERT(!mInFrame);
   MakeCurrent();
   IntRect rect = IntRect(IntPoint(0, 0), GetBufferSize().ToUnknownSize());
-  mCompositor->BeginFrameForWindow(nsIntRegion(rect), Nothing(), rect,
-                                   nsIntRegion());
+  if (!mCompositor->BeginFrameForWindow(nsIntRegion(rect), Nothing(), rect,
+                                        nsIntRegion())) {
+    return false;
+  }
+  mInFrame = true;
   return true;
 }
 
 void RenderCompositorD3D11SWGL::CancelFrame() {
-  mFrameSurfaces.Clear();
+  MOZ_ASSERT(mInFrame);
   mCompositor->CancelFrame();
+  mInFrame = false;
 }
 
 void RenderCompositorD3D11SWGL::CompositorEndFrame() {
-  for (auto& frameSurface : mFrameSurfaces) {
+  nsTArray<FrameSurface> frameSurfaces = std::move(mFrameSurfaces);
+
+  if (!mInFrame) {
+    return;
+  }
+
+  for (auto& frameSurface : frameSurfaces) {
     auto surfaceCursor = mSurfaces.find(frameSurface.mId);
     MOZ_RELEASE_ASSERT(surfaceCursor != mSurfaces.end());
     Surface& surface = surfaceCursor->second;
@@ -161,11 +172,12 @@
       }
     }
   }
-  mFrameSurfaces.Clear();
 }
 
 RenderedFrameId RenderCompositorD3D11SWGL::EndFrame(
     const nsTArray<DeviceIntRect>& aDirtyRects) {
+  MOZ_ASSERT(mInFrame);
+  mInFrame = false;
   mCompositor->EndFrame();
   return GetNextRenderFrameId();
 }